### PR TITLE
{ai}[foss/2021b] PyTorch v1.12.1 w/ Python 3.9.6 w/ CUDA 11.5.2 & dependencies

### PR DESCRIPTION
**Pull request by Flamefire**
(created using `eb --new-pr`)

Update of #17154 using CUDA 11.5 due to an incompatibility.

As this works while the other does not this   
Closes #17154

- [x] Requires #17289

### CODE REVIEW

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 5 out of 6 (6 easyconfigs in total)
taurusi8009 - Linux CentOS Linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor (zen2), 8 x NVIDIA NVIDIA A100-SXM4-40GB, 470.57.02, Python 2.7.5
See https://gist.github.com/fa363fc83c2f8fd9267eaddad8b20a1b for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 2 out of 6 (6 easyconfigs in total)
taurusml24 - Linux RHEL 7.6, POWER, 8335-GTX (power9le), 6 x NVIDIA Tesla V100-SXM2-32GB, 440.64.00, Python 2.7.5
See https://gist.github.com/52599c7ab88241e559b71f2b6bbe9663 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 6 out of 6 (6 easyconfigs in total)
taurusa12 - Linux CentOS Linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz (broadwell), 3 x NVIDIA GeForce GTX 1080 Ti, 460.32.03, Python 2.7.5
See https://gist.github.com/e3cabd5668e5c2ecf4b35a8e4a80112c for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 5 out of 6 (6 easyconfigs in total)
bear-pg0103u14a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz (icelake), 2 x NVIDIA NVIDIA A30, 520.61.05, Python 3.6.8
See https://gist.github.com/7c637f039ff49f68749755c511119bd5 for a full test report.

**Comment by Flamefire** [comment]
@branfosj I added a skip for that failing test similar to that in PyTorch 1.12 for CUDA 11.7

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 5 out of 6 (6 easyconfigs in total)
taurusml5 - Linux RHEL 7.6, POWER, 8335-GTX (power9le), 6 x NVIDIA Tesla V100-SXM2-32GB, 440.64.00, Python 2.7.5
See https://gist.github.com/c81d75ad4be0502437217823c4609154 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (6 easyconfigs in total)
bear-pg0103u14a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz (icelake), 2 x NVIDIA NVIDIA A30, 520.61.05, Python 3.6.8
See https://gist.github.com/139b292b01bb3b2a4eb61cccca13003b for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 81 out of 82 (6 easyconfigs in total)
bear-pg0212u17a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), 1 x NVIDIA Tesla P100-PCIE-16GB, 520.61.05, Python 3.6.8
See https://gist.github.com/595fb693836512854f8aa6ec3b42a8a3 for a full test report.

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 6 out of 6 (6 easyconfigs in total)
gcn5.local.snellius.surf.nl - Linux Rocky Linux 8.7, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz, 1 x NVIDIA NVIDIA A100-SXM4-40GB, 515.86.01, Python 3.6.8
See https://gist.github.com/f337586642709534ed2c3b087594a63b for a full test report.

**Comment by casparvl** [comment]
It would be nice to get this merged... @branfosj @Flamefire I see two remaining test failures in your test reports. Any chance you can investigate? Do we know if these are just flaky test results, or do they point to 'real' issues? If it's only a flaky test, maybe we can just add skips for them. Or we just merge it 'as is' - an easyconfig that works for everything except one test failure, is still better than not having any EasyConfig at all. And we're taking way too long to get this PyTorch stuff merged... :(

**Comment by Flamefire** [comment]
> It would be nice to get this merged... @branfosj @Flamefire I see two remaining test failures in your test reports. Any chance you can investigate? Do we know if these are just flaky test results, or do they point to 'real' issues?

Sorry not right now or in the next ~8-10 weeks.

Feel free to merge this but create an issue to investigate this assigning me so I can look at it when I'm back.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+excluded_tests = {
+    '': [
+        # This test seems to take too long on NVIDIA Ampere at least.
+        'distributed/test_distributed_spawn',
+        # Broken on CUDA 11.5+(?): https://github.com/pytorch/pytorch/issues/75375
+        'distributions/test_constraints',
+        # Those 2 abort on some machines. Skip for now
+        'distributed/fsdp/test_fsdp_input',
+        'distributed/fsdp/test_fsdp_mixed_precision',
+    ]
```
Comment (branfosj) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb:144]:
```suggestion
        # failing on broadwell
        'test_native_mha'
    ]
```

**Comment by casparvl** [comment]
Discussed on Slack with @branfosj. Proposed action:
Disable `distributed/fsdp/test_fsdp_core` and `test_native_mha`, and create at issues for each.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+excluded_tests = {
+    '': [
+        # This test seems to take too long on NVIDIA Ampere at least.
+        'distributed/test_distributed_spawn',
+        # Broken on CUDA 11.5+(?): https://github.com/pytorch/pytorch/issues/75375
+        'distributions/test_constraints',
+        # Those 2 abort on some machines. Skip for now
+        'distributed/fsdp/test_fsdp_input',
+        'distributed/fsdp/test_fsdp_mixed_precision',
```
Comment (casparvl) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb:143]:
```suggestion
        'distributed/fsdp/test_fsdp_mixed_precision',
        # Produces a single failure on some systems
        # See https://github.com/easybuilders/easybuild-easyconfigs/issues/17615
        'distributed/fsdp/test_fsdp_core',
```

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        # failing on broadwell
```
Comment (casparvl) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.12.1-foss-2021b-CUDA-11.5.2.eb:144]:
```suggestion
        # failing on broadwell
        # See https://github.com/easybuilders/easybuild-easyconfigs/issues/17615
```

**Review by casparvl — CHANGES_REQUESTED** [review]

**Comment by casparvl** [comment]
@boegelbot please test @ jsc-zen2

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 6 out of 6 (6 easyconfigs in total)
bear-pg0103u11a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz (icelake), 1 x NVIDIA NVIDIA A100-PCIE-40GB, 520.61.05, Python 3.6.8
See https://gist.github.com/branfosj/def32bc7b2a6d3b439c6bfd722ab066a for a full test report.

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 19 out of 19 (6 easyconfigs in total)
gcn50.local.snellius.surf.nl - Linux Rocky Linux 8.7, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz, 1 x NVIDIA NVIDIA A100-SXM4-40GB, 515.86.01, Python 3.6.8
See https://gist.github.com/casparvl/92ffdef5e61765630592c1b2c059cee0 for a full test report.

**Review by branfosj — APPROVED** [review]

**Comment by casparvl** [comment]
Going in, thanks @Flamefire!

**Comment by casparvl** [comment]
Going in, thanks @Flamefire!

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 5 out of 6 (6 easyconfigs in total)
node3900.accelgor.os - Linux RHEL 8.6, x86_64, AMD EPYC 7413 24-Core Processor (zen3), 1 x NVIDIA NVIDIA A100-SXM4-80GB, 525.85.12, Python 3.6.8
See https://gist.github.com/boegel/202f911923ab0613c6d759e9cee84acb for a full test report.

**Pull request merged**